src/main/java/jnr/ffi/Platform.java

```diff
@@ -474,43 +467,91 @@ public boolean accept(File dir, String name) {
                 exclude = Pattern.compile(".*(lib[a-z]*64|amd64|x86_64).*");
             }
 
-            List<File> matches = new LinkedList<File>();
-            for (String path : libraryPath) {
+            final Pattern versionedLibPattern = Pattern.compile("lib" + libName + "\\.so((?:\\.[0-9]+)*)$");
+
+            FilenameFilter filter = new FilenameFilter() {
+                public boolean accept(File dir, String name) {
+                    return versionedLibPattern.matcher(name).matches();
+                }
+            };
+
+            Map<String, int[]> matches = new HashMap<String, int[]>();
+            for (String path : libraryPaths) {
                 if (exclude.matcher(path).matches()) {
                     continue;
                 }
-                File[] files = new File(path).listFiles(filter);
-                if (files != null && files.length > 0) {
-                    matches.addAll(Arrays.asList(files));
+
+                File libraryPath = new File(path);
+                File[] files = libraryPath.listFiles(filter);
+                if (files == null) {
+                    continue;
+                }
+
+                for (File file : files) {
+                    Matcher matcher = versionedLibPattern.matcher(file.getName());
+                    String versionString = matcher.matches() ? matcher.group(1) : "";
+                    int[] version;
+                    if (versionString == null || versionString.isEmpty()) {
+                        version = new int[0];
+                    } else {
+                        String[] parts = versionString.split("\\.");
+                        version = new int[parts.length - 1];
+                        for (int i = 1; i < parts.length; i++) {
+                            version[i - 1] = Integer.parseInt(parts[i]);
+                        }
+                    }
+                    matches.put(file.getAbsolutePath(), version);
```

Comment:
Simple fix is to use a LinkedHashMap instead of a HashMap for the matches Map. That way insertion order will be retained and the original behaviour should be restored. The only caveat is that, continuing on your example, if `lib` contains a more precies version than `lib64` that the `lib` version will get picked.
Ideally the library loading code should filter out libraries that are not using the correct ABI I guess. That's new functionality that the old code did not do either though.